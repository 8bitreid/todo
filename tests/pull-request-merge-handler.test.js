const pullRequestClosed = require('./fixtures/payloads/pull_request.closed.json')
const { gimmeApp, loadConfig, loadDiff } = require('./helpers')

describe('pull-request-merged-handler', () => {
  let app, github
  const event = { name: 'pull_request', payload: pullRequestClosed }

  beforeEach(() => {
    const gimme = gimmeApp()
    app = gimme.app
    github = gimme.github
  })

  it('creates an issue', async () => {
    await app.receive(event)
    expect(github.issues.create).toHaveBeenCalledTimes(1)
    expect(github.issues.create.mock.calls[0]).toMatchSnapshot()
  })

  it('creates an issue with a truncated title', async () => {
    github.pullRequests.get.mockReturnValueOnce(loadDiff('long-title'))
    await app.receive(event)
    expect(github.issues.create).toHaveBeenCalledTimes(1)
    expect(github.issues.create.mock.calls[0]).toMatchSnapshot()
  })

  it('creates an issue without assigning anyone', async () => {
    github.repos.getContent.mockReturnValueOnce(loadConfig('autoAssignFalse'))
    await app.receive(event)
    expect(github.issues.create.mock.calls[0]).toMatchSnapshot()
  })

  it('creates an issue and assigns the configured user', async () => {
    github.repos.getContent.mockReturnValueOnce(loadConfig('autoAssignString'))
    await app.receive(event)
    expect(github.issues.create.mock.calls[0]).toMatchSnapshot()
  })

  it('creates an issue and assigns the configured users', async () => {
    github.repos.getContent.mockReturnValueOnce(loadConfig('autoAssignArr'))
    await app.receive(event)
    expect(github.issues.create.mock.calls[0]).toMatchSnapshot()
  })

  it('does not create any issues if no todos are found', async () => {
    github.pullRequests.get.mockReturnValueOnce(loadDiff('none'))
    await app.receive(event)
    expect(github.issues.create).toHaveBeenCalledTimes(0)
  })

  it('does not create an issue that already exists', async () => {
    github.search.issues.mockReturnValueOnce(Promise.resolve({
      data: { total_count: 1, items: [{ title: 'I am an example title', state: 'open' }] }
    }))
    await app.receive(event)
    expect(github.issues.create).toHaveBeenCalledTimes(0)
  })

  it('creates many (5) issues', async () => {
    github.pullRequests.get.mockReturnValueOnce(loadDiff('many'))
    await app.receive(event)
    expect(github.issues.create).toHaveBeenCalledTimes(5)
  })

  it('ignores changes to the config file', async () => {
    github.pullRequests.get.mockReturnValueOnce(loadDiff('config'))
    await app.receive(event)
    expect(github.issues.create).toHaveBeenCalledTimes(0)
  })

  it('ignores changes to the bin directory', async () => {
    github.pullRequests.get.mockReturnValueOnce(loadDiff('bin'))
    github.repos.getContent.mockReturnValueOnce(loadConfig('excludeBin'))
    await app.receive(event)
    expect(github.issues.createComment).toHaveBeenCalledTimes(0)
  })

  it('creates an issue with a body line', async () => {
    github.pullRequests.get.mockReturnValueOnce(loadDiff('body'))
    await app.receive(event)
    expect(github.issues.create.mock.calls[0]).toMatchSnapshot()
  })

  it('reopens a closed issue', async () => {
    github.search.issues.mockReturnValueOnce(Promise.resolve({
      data: { total_count: 1, items: [{ title: 'I am an example title', state: 'closed' }] }
    }))
    await app.receive(event)
    expect(github.issues.edit).toHaveBeenCalledTimes(1)
    expect(github.issues.createComment).toHaveBeenCalledTimes(1)
    expect(github.issues.createComment.mock.calls[0]).toMatchSnapshot()
    expect(github.issues.create).toHaveBeenCalledTimes(0)
  })

  it('respects the reopenClosed config', async () => {
    github.repos.getContent.mockReturnValueOnce(loadConfig('reopenClosedFalse'))
    github.search.issues.mockReturnValueOnce(Promise.resolve({
      data: { total_count: 1, items: [{ title: 'I am an example title', state: 'closed' }] }
    }))
    await app.receive(event)
    expect(github.issues.edit).toHaveBeenCalledTimes(0)
    expect(github.issues.createComment).toHaveBeenCalledTimes(0)
    expect(github.issues.create).toHaveBeenCalledTimes(0)
  })
})
