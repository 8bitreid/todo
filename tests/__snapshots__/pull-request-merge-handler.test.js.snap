// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`pull-request-merged-handler creates an issue 1`] = `
Array [
  Object {
    "assignee": "JasonEtco",
    "body": "https://github.com/JasonEtco/tests/blob/6c2373ad036d292f7e9467981ba8b467e8b8b0ff/.travis.yml#L11

---

###### This issue was generated by [todo](https://todo.jasonet.co) based on a \`TODO\` comment in 6c2373ad036d292f7e9467981ba8b467e8b8b0ff. It&#x27;s been assigned to @JasonEtco because they committed the code.",
    "labels": Array [
      "todo",
    ],
    "owner": "JasonEtco",
    "repo": "tests",
    "title": "I am an example title",
  },
]
`;

exports[`pull-request-merged-handler creates an issue and assigns the configured user 1`] = `
Array [
  Object {
    "assignee": "matchai",
    "body": "https://github.com/JasonEtco/tests/blob/6c2373ad036d292f7e9467981ba8b467e8b8b0ff/.travis.yml#L11

---

###### This issue was generated by [todo](https://todo.jasonet.co) based on a \`TODO\` comment in 6c2373ad036d292f7e9467981ba8b467e8b8b0ff. It&#x27;s been automagically assigned to @matchai.",
    "labels": Array [
      "todo",
    ],
    "owner": "JasonEtco",
    "repo": "tests",
    "title": "I am an example title",
  },
]
`;

exports[`pull-request-merged-handler creates an issue and assigns the configured users 1`] = `
Array [
  Object {
    "assignees": Array [
      "JasonEtco",
      "matchai",
      "defunkt",
    ],
    "body": "https://github.com/JasonEtco/tests/blob/6c2373ad036d292f7e9467981ba8b467e8b8b0ff/.travis.yml#L11

---

###### This issue was generated by [todo](https://todo.jasonet.co) based on a \`TODO\` comment in 6c2373ad036d292f7e9467981ba8b467e8b8b0ff. It&#x27;s been automagically assigned to @JasonEtco, @matchai and @defunkt.",
    "labels": Array [
      "todo",
    ],
    "owner": "JasonEtco",
    "repo": "tests",
    "title": "I am an example title",
  },
]
`;

exports[`pull-request-merged-handler creates an issue with a truncated title 1`] = `
Array [
  Object {
    "assignee": "JasonEtco",
    "body": "https://github.com/JasonEtco/tests/blob/6c2373ad036d292f7e9467981ba8b467e8b8b0ff/.travis.yml#L11

---

###### This issue was generated by [todo](https://todo.jasonet.co) based on a \`TODO\` comment in 6c2373ad036d292f7e9467981ba8b467e8b8b0ff. It&#x27;s been assigned to @JasonEtco because they committed the code.",
    "labels": Array [
      "todo",
    ],
    "owner": "JasonEtco",
    "repo": "tests",
    "title": "This is a really, really, really, really, really, really, really, really, really...",
  },
]
`;

exports[`pull-request-merged-handler creates an issue without assigning anyone 1`] = `
Array [
  Object {
    "body": "https://github.com/JasonEtco/tests/blob/6c2373ad036d292f7e9467981ba8b467e8b8b0ff/.travis.yml#L11

---

###### This issue was generated by [todo](https://todo.jasonet.co) based on a \`TODO\` comment in 6c2373ad036d292f7e9467981ba8b467e8b8b0ff.",
    "labels": Array [
      "todo",
    ],
    "owner": "JasonEtco",
    "repo": "tests",
    "title": "I am an example title",
  },
]
`;
